File: src/conflict/conflict.service.ts
import { HttpStatus, Injectable } from '@nestjs/common';
import { User } from '@prisma/client';
import { Prisma } from '@prisma/client';
import { AuthService } from '../auth/auth.service';
import { ok } from '../common/dto/api-response.dto';
import { DomainException } from '../common/errors/domain.exception';
import { findActiveRoomById, requireActiveRoomMember } from '../common/room-access.util';
import { BaseSoftDeleteService } from '../common/soft-delete/base-soft-delete.service';
import { ACTIVE_DEL_YN } from '../common/soft-delete/soft-delete.util';
import { ConsensusService } from '../consensus/consensus.service';
import { PrismaService } from '../prisma/prisma.service';

@Injectable()
export class ConflictService extends BaseSoftDeleteService {
  constructor(
    private readonly authService: AuthService,
    private readonly consensusService: ConsensusService,
    private readonly prisma: PrismaService,
  ) {
    super();
  }

  private async requireRoomMember(roomId: bigint, userId: bigint) {
    await requireActiveRoomMember({
      prisma: this.prisma,
      activeWhere: this.activeWhere.bind(this),
      roomId,
      userId,
    });
  }

  async getConflictMap(roomId: number, user: User) {
    await this.requireRoomMember(BigInt(roomId), user.id);

    const room = await findActiveRoomById({
      prisma: this.prisma,
      activeWhere: this.activeWhere.bind(this),
      roomId: BigInt(roomId),
      include: {
        memberProfiles: {
          where: this.activeWhere({}),
          include: { user: true },
          orderBy: { createdAt: 'asc' },
        },
      },
    });
    if (room.memberProfiles.length < 2) {
      throw new DomainException(HttpStatus.UNPROCESSABLE_ENTITY, 'ROOM_NOT_READY', '갈등 지도를 계산할 프로필이 부족합니다.');
    }

    const members = room.memberProfiles.map((profile, index) => ({
      userId: Number(profile.userId),
      nickname: profile.user.nickname,
      joinedOrder: index,
      scores: {
        mobility: profile.mobilityScore,
        photo: profile.photoScore,
        budget: profile.budgetScore,
        theme: profile.themeScore,
      },
    }));

    const analysis = this.consensusService.analyzeGroup(members);
    const topConflict = analysis.conflictAxes[0];
    const highMember = topConflict ? members.find((member) => member.userId === topConflict.highUserId) : null;
    const lowMember = topConflict ? members.find((member) => member.userId === topConflict.lowUserId) : null;
    const summaryText = topConflict
      ? `${highMember?.nickname ?? 'A님'}과 ${lowMember?.nickname ?? 'B님'}은 ${this.axisLabel(topConflict.axis)}에서 ${topConflict.gap}점 차이로 충돌합니다.`
      : '현재 그룹은 공통 지대가 넓습니다.';

    const conflictMap = await this.prisma.conflictMap.create({
      data: {
        roomId: room.id,
        commonAxes: analysis.commonAxes,
        conflictAxes: analysis.conflictAxes as unknown as Prisma.InputJsonValue,
        summaryText,
        delYn: ACTIVE_DEL_YN,
      },
    });

    return ok({
      roomId,
      conflictMapId: Number(conflictMap.id),
      commonAxes: analysis.commonAxes,
      conflictAxes: analysis.conflictAxes.map(({ axis, gap, severity }) => ({ axis, gap, severity })),
      summaryText,
      members: members.map((member) => ({
        userId: member.userId,
        nickname: member.nickname,
        scores: member.scores,
      })),
    });
  }

  private axisLabel(axis: string) {
    switch (axis) {
      case 'mobility':
        return '활동성';
      case 'photo':
        return '기록';
      case 'budget':
        return '예산';
      case 'theme':
        return '테마';
      default:
        return axis;
    }
  }
}

File: src/main/kotlin/com/tripsync/application/conflict/ConflictService.kt
package com.tripsync.application.conflict

import com.tripsync.application.consensus.ConsensusService
import com.tripsync.application.consensus.MemberSnapshot
import com.tripsync.common.dto.ApiResponse
import com.tripsync.common.exception.DomainException
import com.tripsync.domain.entity.User
import com.tripsync.domain.enums.YnFlag
import com.tripsync.domain.repository.*
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class ConflictService(
    private val consensusService: ConsensusService,
    private val tripRoomRepository: TripRoomRepository,
    private val roomMemberRepository: RoomMemberRepository,
    private val roomMemberProfileRepository: RoomMemberProfileRepository,
    private val conflictMapRepository: ConflictMapRepository,
) {

    @Transactional
    fun getConflictMap(roomId: Long, user: User): ApiResponse<Map<String, Any>> {
        validateRoomMember(roomId, user.id)

        val room = tripRoomRepository.findById(roomId)
            .orElseThrow { DomainException(HttpStatus.NOT_FOUND, "ROOM_NOT_FOUND", "존재하지 않는 방입니다.") }

        val profiles = roomMemberProfileRepository.findAllByRoomIdAndDelYn(roomId, YnFlag.N)
        if (profiles.size < 2) {
            throw DomainException(HttpStatus.UNPROCESSABLE_ENTITY, "ROOM_NOT_READY", "갈등 지도를 계산할 프로필이 부족합니다.")
        }

        val members = profiles.mapIndexed { index, profile ->
            MemberSnapshot(
                userId = profile.user.id,
                nickname = profile.user.nickname,
                scores = com.tripsync.domain.entity.AxisScores(
                    mobility = profile.mobilityScore,
                    photo = profile.photoScore,
                    budget = profile.budgetScore,
                    theme = profile.themeScore,
                ),
                joinedOrder = index,
            )
        }

        val analysis = consensusService.analyzeGroup(members)
        val topConflict = analysis.conflictAxes.firstOrNull()
        val highMember = topConflict?.let { members.find { m -> m.userId == it.highUserId } }
        val lowMember = topConflict?.let { members.find { m -> m.userId == it.lowUserId } }
        val summaryText = topConflict?.let {
            "${highMember?.nickname ?: "A님"}과 ${lowMember?.nickname ?: "B님"}은 ${axisLabel(it.axis)}에서 ${it.gap}점 차이로 충돌합니다."
        } ?: "현재 그룹은 공통 지대가 넓습니다."

        val conflictMap = conflictMapRepository.save(
            com.tripsync.domain.entity.ConflictMap(
                room = room,
                commonAxes = analysis.commonAxes.map { it.name.lowercase() },
                conflictAxes = analysis.conflictAxes.map {
                    mapOf(
                        "axis" to it.axis.name.lowercase(),
                        "min" to it.min,
                        "max" to it.max,
                        "gap" to it.gap,
                        "severity" to it.severity.name.lowercase(),
                        "highUserId" to it.highUserId,
                        "lowUserId" to it.lowUserId,
                    )
                },
                summaryText = summaryText,
            )
        )

        return ApiResponse.ok(
            mapOf(
                "roomId" to roomId,
                "conflictMapId" to conflictMap.id,
                "commonAxes" to analysis.commonAxes.map { it.name.lowercase() },
                "conflictAxes" to analysis.conflictAxes.map {
                    mapOf(
                        "axis" to it.axis.name.lowercase(),
                        "gap" to it.gap,
                        "severity" to it.severity.name.lowercase(),
                    )
                },
                "summaryText" to summaryText,
                "members" to members.map {
                    mapOf(
                        "userId" to it.userId,
                        "nickname" to it.nickname,
                        "scores" to mapOf(
                            "mobility" to it.scores.mobility,
                            "photo" to it.scores.photo,
                            "budget" to it.scores.budget,
                            "theme" to it.scores.theme,
                        ),
                    )
                },
            )
        )
    }

    private fun validateRoomMember(roomId: Long, userId: Long) {
        if (!roomMemberRepository.existsByRoomIdAndUserIdAndDelYn(roomId, userId, YnFlag.N)) {
            throw DomainException(HttpStatus.FORBIDDEN, "FORBIDDEN", "방 멤버만 접근할 수 있습니다.")
        }
    }

    private fun axisLabel(axis: com.tripsync.domain.enums.ScoreAxis): String {
        return when (axis) {
            com.tripsync.domain.enums.ScoreAxis.MOBILITY -> "활동성"
            com.tripsync.domain.enums.ScoreAxis.PHOTO -> "기록"
            com.tripsync.domain.enums.ScoreAxis.BUDGET -> "예산"
            com.tripsync.domain.enums.ScoreAxis.THEME -> "테마"
        }
    }
}
